spikeinterface/sortingcomponents/peak_detection.py

```diff
@@ -347,3 +405,152 @@ def _numba_detect_peak_neg(traces, traces_center, peak_mask, exclude_sweep_size,
+        # -- torch argrelmin
+        if peak_sign == "neg":
+            neg_traces = torch.as_tensor(
+                -traces, device=device, dtype=torch.float
+            )
+        elif peak_sign == "pos":
+            neg_traces = torch.as_tensor(
+                traces, device=device, dtype=torch.float
+            )
+        elif peak_sign == "both":
+            neg_traces = torch.as_tensor(
+                -np.abs(traces), device=device, dtype=torch.float
+            )
+        thresholds_torch = torch.as_tensor(thresholds, device=device, dtype=torch.float)
+
+        max_amps, inds = F.max_pool2d_with_indices(
+            neg_traces[None, None],
+            kernel_size=[2 * exclude_sweep_size + 1, 1],
+            stride=1,
+            padding=[exclude_sweep_size, 0],
+        )
+        max_amps = max_amps[0, 0]
+        inds = inds[0, 0]
+        # torch `inds` gives loc of argmax at each position
+        # find those which actually *were* the max
+        unique_inds = inds.unique()
+        window_max_inds = unique_inds[inds.view(-1)[unique_inds] == unique_inds]
+
+        # voltage threshold (@Charlie, added by channel)
+        thresholds_by_channel = torch.as_tensor(
+                thresholds_torch[window_max_inds % num_channels], device=device, dtype=torch.float
+            )
+        max_amps_at_inds = max_amps.view(-1)[window_max_inds]
+        crossings = torch.nonzero(max_amps_at_inds > thresholds_by_channel).squeeze()
+        if not crossings.numel():
+            return np.array([]), np.array([]), np.array([])
+
+        # -- unravel the spike index
+        # (right now the indices are into flattened recording)
+        peak_inds = window_max_inds[crossings]
+        sample_inds = torch.div(peak_inds, num_channels, rounding_mode="floor")
+        chan_inds = peak_inds % num_channels
+        amplitudes = max_amps_at_inds[crossings]
+
+        # we need this due to the padding in convolution
+        valid_inds = torch.nonzero(
+            (0 < sample_inds) & (sample_inds < traces.shape[0] - 1)
+        ).squeeze()
+        if not sample_inds.numel():
+            return np.array([]), np.array([]), np.array([])
+        sample_inds = sample_inds[valid_inds]
+        chan_inds = chan_inds[valid_inds]
+        amplitudes = amplitudes[valid_inds]
+
+        # -- deduplication
+        # We deduplicate if the channel index is provided.
+        if neighbours_mask is not None:
+            neighbours_mask = torch.tensor(
+                neighbours_mask, device=device, dtype=torch.long
+            )
+
+            # -- temporal max pool
+            # still not sure why we can't just use `max_amps` instead of making
+            # this sparsely populated array, but it leads to a different result.
+            max_amps[:] = 0
+            max_amps[sample_inds, chan_inds] = amplitudes
+
+            # TODO: @Charlie
+            # 1. is this to deduplicate over neighbours using a wider window? Assuming propagation
```

Comment:
Yes, this is a temporal deduplication. Actually, this is a great point -- this might not be necessary, maybe we can combine this into the previous max pool...